apps/rush-lib/src/cli/actions/PublishAction.ts

```diff
@@ -487,7 +488,7 @@ export class PublishAction extends BaseRushAction {
     }
 
     if (this._npmAuthToken.value) {
-      args.push(`--${registry}:_authToken=${this._npmAuthToken.value}`);
+      args.push(`--${registry}:_authToken=\$\{${this._npmAuthToken.environmentVariable}\}`);
```

Comment:
This is still going to show up in the build log I think.  

@iclanton  Do you know if there is a way to pass tokens to PNPM/NPM via an environment variable, instead of on the command-line or npmrc file?

Reply:
I was unsure about this too. My VS CODE output while debugging publish was: `--//registry.npmjs.org/:_authToken=${NPM_TOKEN} --access public (apps\rush-publish-test)`. I'm not using a local .npmrc file, and just have NPM_TOKEN set in my machine's environment variables. I have tested this multiple times, but, relying on the output via VS code. I'm not sure on how other systems would handle it.

Reply:
Yeah you're right. The current code only allowed me to publish using the environmentVariable property, it shouldnt be visible. Id love to hear more about this if anyone has some information, ill update this as soon as possible.

Reply:
[These instructions](https://blog.npmjs.org/post/118393368555/deploying-with-npm-private-modules) and [this thread](https://npm.community/t/cannot-set-npm-config-keys-containing-underscores-registry-auth-tokens-for-example-via-npm-config-environment-variables/233/13) suggest that you have to invent your own environment variable name, and put it in an `.npmrc` file, like this:

**.npmrc**
```
//registry.npmjs.org/:_authToken=${MY_AUTH_TOKEN}
```

Then the CI machine can specify `MY_AUTH_TOKEN` as an environment variable when running `npm publish`.  

However, there is a problem that we probably DON'T want to specify this variable during normal working.  For example, on GitHub for normal development we want to have:

**common/config/rush/.npmrc**
```
registry=https://registry.npmjs.org/
always-auth=false
```

...whereas in a private repo with a private registry, we'd want something like:

**common/config/rush/.npmrc**
```
always-auth=true
```
**~/.npmrc**
```

Reply:
I chatted with @iclanton  about this.  It seems that the line must go in an `.npmrc` file. According to the [npmrc docs](https://docs.npmjs.com/files/npmrc), there are only 4 possible locations for this file.  In order of precedence:

- per-project config file (/path/to/my/project/.npmrc)
- per-user config file (~/.npmrc)
- global config file ($PREFIX/etc/npmrc)
- npm builtin config file (/path/to/npm/npmrc)

We considered writing `/path/to/my/project/.npmrc` before `npm publish` and deleting it afterwards, but that can leave an unclean state if the CI job gets cancelled.  Also, there may be valid cases for user-defined settings in this file location.

So instead we are thinking Rush should use `~/.npmrc`.  The home folder can get redirected fairly easily.  For example, on Windows if you do:

```
$ set USERPROFILE=C:\Git\rushstack\common\temp\publish-home
$ cd C:\Git\rushstack\my\project
$ npm config list
```
...then it will print out settings found in `C:\Git\rushstack\common\temp\publish-home\.npmrc`.

A possible risk is that `npm publish` might somehow depend on other state from the home directory, but this seems unlikely.  Note that we would only be redirecting `USERPROFILE` during that specific command `npm publish`.  The `rush install` and `rush build` stuff would have access to the normal home directory.

Where does `common\temp\publish-home\.npmrc` come from?  We're proposing that the user would create a file `common\config\rush\.npmrc-publish`, and add it to Git.  Rush would copy it to `common\temp\publish-home` prior to `rush publish`.

For backwards compatibility, if `.npmrc-publish` does not exist, then we would fall back to the old behavior.

@darsi-an How does that sound?  It seems like probably not too much work to implement.  The only slightly tricky part is we would probably need separate logic for redirecting the home directory on each OS (Mac/Linux/Windows).  

If you're able to implement+test the file copying idea, I can help with the documentation and `rush init` part.

Reply:
Sounds good, I like the approach. I'll update my PR accordingly. Regarding the `rush init` part, is it to generate a `common\config\rush\.npmrc-publish` file?